### ISSUE TITLE
Resource not available after aws_ram_principal_association

### ISSUE THREAD

**Issue by CliffHoogervorst** [terraform-providers/terraform-provider-aws#11902]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

Terraform 0.12.19
AWS provider 2.33.0
### Affected Resource(s)

* aws_route53_resolver_rule_association
* aws_ram_principal_association

### Terraform Configuration Files
```hcl
# Find the RAM share in the central account
data "aws_ram_resource_share" "route53outbound" {
  name           = "route53-outbound"
  resource_owner = "SELF"
  provider       = aws.route53resolver
}
# Share the resource share with the current account
resource "aws_ram_principal_association" "route53resolver-outbound" {
  provider           = aws.route53resolver
  principal          = data.aws_caller_identity.current.account_id
  resource_share_arn = data.aws_ram_resource_share.route53outbound.arn
}

# Get the route53resolver rule-ids from central account
data "aws_route53_resolver_rules" "sharedrules" {
  provider     = aws.route53resolver
  owner_id     = "11111111111"
  rule_type    = "FORWARD"
  share_status = "SHARED_BY_ME"
}
# Associate route53 rules with current account
resource "aws_route53_resolver_rule_association" "r53rules" {
  for_each         = data.aws_route53_resolver_rules.sharedrules.resolver_rule_ids
  resolver_rule_id = each.value
  vpc_id           = aws_vpc.vpc.id
  name             = "resolver-rule"
  depends_on       = [aws_ram_principal_association.route53resolver-outbound]
}
```

### Expected Behavior

- RAM share shared with target account
- Route53 rules (in the RAM share) associated with VPC in target account

### Actual Behavior

Some rules fail to associate on first run:

```
Error: error creating Route 53 Resolver rule association: ResourceNotFoundException: [RSLVR-00703] Resolver rule with ID "rslvr-rr-a662b14cd47b457e9" does not exist.
	status code: 400, request id: 68ef49b8-d99b-4bc5-ab5f-95dc2fc59f4d

  on route53resolver_rules.tf line 22, in resource "aws_route53_resolver_rule_association" "r53rules":
  22: resource "aws_route53_resolver_rule_association" "r53rules" {
```

### Steps to Reproduce

1. `terraform apply`
Will only fail on the first run. Second run is ok
### Important Factoids

Issue can be reproduced with AWS CLI:
In the central account I run the RAM association:
```aws ram associate-resource-share --resource-share-arn arn:aws:ram:eu-west-1:11111111111:resource-share/cab4b954-16e6-5a53-8e35-d0371837dd1d --principals 44444444444```

Then I watch the status of the association:
```aws ram get-resource-share-associations --resource-share-arn arn:aws:ram:eu-west-1:11111111111:resource-share/cab4b954-16e6-5a53-8e35-d0371837dd1d --principal 44444444444 --association-type PRINCIPAL```

The status very quickly switches from "ASSOCIATING" to "ASSOCIATED".
At the same time I watch the route53 rules in the target account using:
```aws route53resolver list-resolver-rules```

When the status of the RAM share switches to ASSOCIATED not all rules are visible in the target account. The exact number depends on timing and the situation is there for only a few seconds

AWS support confirmed that when RAM share is associated not all resources might be available directly

```
I received an update from our internal team and they confirmed that RAM begins the process of sharing workflows asynchronously when associations reach ASSOCIATED. The resources will be visible in the 'list-resolver-rules' API output only after this sharing workflows finishes. Therefore, there is a delay between the time RAM displays a resource as ASSOCIATED and the resource appears in 'list-resolver-rules' API action. The delay is variable but with more the resolver rules multiplied by the number of accounts it is shared with, it will make the delay worse. For example, if you shared 30 resolver rules with 1 account that should be relatively quick, but if you shared 30 resolver rules with an organization and that org has 100 accounts that will be slower. Also, if you are sharing with 1 account, then it could be due to system's eventual consistency behavior. 
```

We share 27 rules with 100+ accounts in 1 RAM share. 

Seems like Terraform has to "wait" or retry until the resource is really there.

**Comment by bflad** [comment]
This comment/thread is related: https://github.com/terraform-providers/terraform-provider-aws/issues/12729#issuecomment-628277627

**Comment by ewbankkit** [comment]
Closing as a duplicate of #12729.
